ddl/ddl_api.go

```diff
@@ -2245,6 +2253,11 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
 		// nothing to do.
 		return nil
 	}
+	// Not allow to change binary to utf8mb4 or utf8.
+	if origCharset == charset.CharsetBin {
```

Comment:
I think you should put the code into the `modifiableCharsetAndCollation`

Reply:
Put the code into the modifiableCharsetAndCollation need to check the col.Tp.

Reply:
You should check the altering charset only if the original charset is not equal to the changing charset, then the changing charset can’t be binary.